racket/collects/racket/private/generic-methods.rkt

```diff
@@ -172,28 +190,46 @@
 
   (define-syntax (generic-methods stx)
     (syntax-case stx ()
-      [(_ gen combine #:scope scope def ...)
+      [(_ gen combine #:scope scope #:check? check? def ...)
        (let ()
+         (unless (boolean? (syntax-e #'check?))
+           (raise-syntax-error
+            'generic-methods
+            "check? must be a boolean literal"
+            #'check?))
+
          (define info (get-info 'generic-methods stx #'gen))
          (define orig-id (generic-info-name info))
          (define methods (map (make-method-delta #'scope orig-id)
                               (generic-info-method-names info)))
-         (with-syntax ([(method ...) methods])
+         (define checking? (syntax-e #'check?))
+         (define req-methods (map (λ (m)
+                                    (if (and m checking?)
+                                        #'#t
+                                        #'#f))
+                                  (generic-info-required-methods info)))
+         (with-syntax ([(method ...) methods]
+                       [(req-method ...) req-methods])
            (syntax/loc stx
              (syntax-parameterize ([generic-method-outer-context #'gen])
                (letrec-syntaxes+values
-                ([(method) (make-unimplemented 'method)] ...)
-                ()
-                (syntax-parameterize ([generic-method-inner-context #'gen])
-                  def ...
-                  (combine (implementation method) ...)))))))]
+                   ([(method) (make-unimplemented
+                               'method
+                               req-method
+                               (list (quote-syntax def) ...))] ...)
```

Comment:
It looks like this copies all definitions into a `make-unimplemented` for each method. If that's right, I can't help worry about the O(N^2)-ish expansion, even if N is normally small in practice. Maybe share the list, something like the code below?
```
[(method ...) 
 (let ([defs (list (quote-syntax def) ...)])
    (values (make-unimplemented 'method req-method defs) ...))]
```

Reply:
Just curious. Say, if `letrec-syntaxes+values` doesn't support multiple values, would wrapping it with `(let-syntax ([defs (list (quote-syntax def) ...)]) ...)` and then using `(syntax-local-value #'defs)` be a good sharing mechanism?

Reply:
Yes, I think that would work.